packages/loopring_v3/test/testUserStakingPool.ts

```diff
@@ -1,266 +1,307 @@
+import { Artifacts } from "../util/Artifacts";
+import { advanceTimeAndBlockAsync } from "../util/TimeTravel";
 import { expectThrow } from "./expectThrow";
 import BN = require("bn.js");
-import { ExchangeTestUtil } from "./testExchangeUtil";
+const truffleAssert = require("truffle-assertions");
 
 contract("UserStakingPool", (accounts: string[]) => {
-  const owner1 = accounts[0];
-  const owner2 = accounts[1];
-  const owner3 = accounts[2];
-  const emptyAddr = "0x0000000000000000000000000000000000000000";
+  const contracts = new Artifacts(artifacts);
+  const MockContract = contracts.MockContract;
+  const UserStakingPool = contracts.UserStakingPool;
 
-  let userstaking: any;
-  let protocolfee: any;
-  let exchangeTestUtil: ExchangeTestUtil;
+  const ZERO = new BN(0);
+
+  var mockLRC: any;
+  var mockProtocolFeeVault: any;
+  var userStakingPool: any;
+
+  var MIN_WITHDRAW_DELAY: number;
+  var MIN_CLAIM_DELAY: number;
 
   before(async () => {
-    exchangeTestUtil = new ExchangeTestUtil();
-    await exchangeTestUtil.initialize(accounts);
-    userstaking = exchangeTestUtil.userStakingPool;
-    protocolfee = exchangeTestUtil.protocolFeeVaultContract;
-    userstaking.setProtocolFeeVault(protocolfee.address);
+    mockLRC = await MockContract.new();
+    mockProtocolFeeVault = await MockContract.new();
+    userStakingPool = await UserStakingPool.new(mockLRC.address);
```

Comment:
I guess the downside of using a mock address here is that token transfers aren't being tested. You could comment out all token transfers in UserStakingPool and the tests would still pass. So if you make a mistake when calling the token transfer functions these tests will not catch that.

Arguably this could make them more _integration_ tests than _unit_ tests, but I feel like it needs to be tested somewhere, so it may as well be in these tests. I think it makes sense that if you for example test the 'stake' function that you also test that LRC if effectively transferred correctly and staked.

Reply:
The current tests haven't actually mocked the non-default behavior of the depending contracts yet, such as LRC transfer failure. These cases need to be tested here in this file. I'll create another PR for more tests.